File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] lmul = new int[n];
        int[] rmul = new int[n];
        
        lmul[0] = 1;
        for(int i=1; i<n; i++) {
            lmul[i] = lmul[i-1] * nums[i-1];
        }
        rmul[n-1] = 1;
        for(int i=n-2; i>=0; i--) {
            rmul[i] = rmul[i+1] * nums[i+1];
        }
        
        int[] ans = new int[n];
        
        for(int i=0; i<n; i++) {
            ans[i] = lmul[i] * rmul[i];
        }
        return ans;
        
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> leftProduct(nums.size(), 1);
        vector<int> rightProduct(nums.size(), 1);

        for(int i=1; i<nums.size(); i++) {
            leftProduct[i] = nums[i-1] * leftProduct[i-1];
        }

        for(int i=nums.size()-2; i>=0; i--) {
            rightProduct[i] = nums[i+1] * rightProduct[i+1];
        }

        vector<int> ans;
        for(int i=0; i<nums.size(); i++) {
            ans.push_back(leftProduct[i] * rightProduct[i]);
        }
        return ans;
    }
};

















//         int n = nums.size();
//         vector<int> answer(n, 1);
//         for(int i=n-2; i>=0; i--) {
//             answer[i] = answer[i+1] * nums[i+1];
//         }

//         int L = 1;
//         for(int i=0; i<n; i++) {
//             answer[i] = answer[i] * L;
//             L = L * nums[i];
//         }

//         return answer;